### PR TITLE
LibAFL Dynamic Analysis

### PR DESCRIPTION
**Pull request by tokatoka**
follow up for #2178 
the plan is 
- [x] this pass logs all the hashed function signature.
- [x] and writes the hash into a global, this list should contain all the hashes of the functions that executions went through
- [x] an observer (built upon ListObserver) should observe the list 
- [x] then do the lookup to find the static analysis info

### CODE REVIEW

**Comment by tokatoka** [comment]
@domenukk 
done!

File: fuzzers/dynamic_analysis/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+debug = false
+strip = true
+
+[build-dependencies]
+cc = { version = "1.0", features = ["parallel"] }
+which = "4.4"
+
+[dependencies]
+env_logger = "0.10"
+once_cell = "1.19"
```
Comment (domenukk) [inline fuzzers/dynamic_analysis/Cargo.toml:29]: once_cell is in the standard library now, no need to use the trait anymore

File: libafl_cc/build.rs [hunk 2]
```diff
@@ -383,6 +383,7 @@ pub const LIBAFL_CC_LLVM_VERSION: Option<usize> = None;
     if cfg!(unix) {
         cxxflags.push(String::from("-shared"));
         cxxflags.push(String::from("-fPIC"));
+        cxxflags.push(String::from("-g"));
```
Comment (domenukk) [inline libafl_cc/build.rs:386]: Are there never any downsides to `-g`?

File: libafl_targets/Cargo.toml [hunk 3]
```diff
@@ -68,6 +68,8 @@ rustversion = "1.0"
 libafl = { path = "../libafl", version = "0.12.0", default-features = false, features = [] }
 libafl_bolts = { path = "../libafl_bolts", version = "0.12.0", default-features = false, features = [] }
 libc = "0.2"
+hashbrown = "0.14"
+once_cell = "1.19"
```
Comment (domenukk) [inline libafl_targets/Cargo.toml:72]: same here, not needed usually

File: libafl/src/observers/profiler.rs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+use core::fmt::Debug;
+use std::path::Path;
+use alloc::string::String;
+
+use libafl_bolts::{ownedref::OwnedMutPtr, Error, Named};
+use serde::{Deserialize, Serialize};
+
+use crate::{inputs::UsesInput, observers::Observer};
+
+use std::collections::HashMap;
```
Comment (domenukk) [inline libafl/src/observers/profiler.rs:14]: that's the wrong hashmap?

File: libafl_cc/build.rs (hunk 2 above)
Comment (tokatoka) [reply libafl_cc/build.rs:386]: was for debug.
forgot to delete it

File: libafl_targets/Cargo.toml [hunk 5]
```diff
@@ -68,6 +68,8 @@ rustversion = "1.0"
 libafl = { path = "../libafl", version = "0.12.0", default-features = false, features = [] }
 libafl_bolts = { path = "../libafl_bolts", version = "0.12.0", default-features = false, features = [] }
 libc = "0.2"
+hashbrown = "0.14"
```
Comment (domenukk) [inline libafl_targets/Cargo.toml:71]: Maybe hide the new feature behind a flag and make this dependency optional?

File: fuzzers/dynamic_analysis/Cargo.toml (hunk 1 above)
Comment (tokatoka) [reply fuzzers/dynamic_analysis/Cargo.toml:29]: the one i want to use LazyCell is nightly only

https://doc.rust-lang.org/std/cell/struct.LazyCell.html

**Pull request merged**